### ISSUE TITLE
Cannot use JRE.xcframework with Xcode 15 Release Candidate due to code signing issue

### ISSUE THREAD

**Issue by mungler** [google/j2objc#2169]
Hi,

We've been using JRE.xcframework for a while now, and its been working great. We note the imminent release of macOS Sonoma and Xcode 15, so decided to try building in the Xcode 15 RC. However, our app now refuses to build, with the error: 

"<app path>/frameworks/JRE.xcframework:1:1 The signature of “JRE.xcframework” cannot be verified", and when clicked, offers to move it to the trash, with a further message:

"Validation Error: bundle format unrecognized, invalid, or unsuitable"

I've tried:

- changing JRE.xcframework to 'Embed & Sign' in the 'Frameworks, Libraries and Embedded Content' section of the 'General' tab of the build settings. 
- code signing JRE.xcframework manually via the `codesign` tool, but it gives me the same 'unrecognized, invalid, or unsuitable' message.
- code signing the individual libjre_emul.a files inside the various platform-specific subfolders

None of these seem to help.

I'm wondering if there's something missing from the generated xcframework bundles - i've seen reference to a Resources folder, is there perhaps some file that needs to live there?

Anyone with a better handle on this stuff able to help?

**Comment by mungler** [comment]
Closing in favour of #2101 - sorry, failed to notice this has been logged already. This will prevent us upgrading to Sonoma when its released later this month until its resolved.

**Comment by bryan1anderson** [comment]
You included a couple steps I hadn't tried yet. Thanks for this. Also have run into the same problem..

**Comment by litstrong** [comment]
Hi mungler@, may I ask where you see the references to the Resources folder?

**Comment by mungler** [comment]
Re: "Resources", i'll try to find out, sorry, I just noticed in passing when I spent most of yesterday googling on the subject. Will update here if/when I find anything concrete.

Here's the thing that makes me think the package is malformed somehow - even reading the signing info results in an error. Compare:

```
rory@Zaphod ~> touch foo
rory@Zaphod ~> codesign -dv foo
foo: code object is not signed at all

rory@Zaphod ~ [1]> codesign --timestamp -s "Apple Development: Rory Sinclair (99******HA)" foo
rory@Zaphod ~> codesign -dv foo
Executable=/Users/rory/foo
Identifier=foo
Format=generic
CodeDirectory v=20200 size=131 flags=0x0(none) hashes=0+2 location=embedded
Signature size=9172
Timestamp=14 Sep 2023 at 09:24:29
Info.plist=not bound
TeamIdentifier=B********5
Sealed Resources=none
Internal requirements count=1 size=168
rory@Zaphod ~> 
```

versus:

```
rory@Zaphod ~/W/j/d/frameworks (master)> codesign -dv JRE.xcframework
JRE.xcframework: bundle format unrecognized, invalid, or unsuitable
rory@Zaphod ~/W/j/d/frameworks (master) [1]> 
```

So it seems the codesign tool doesn't even recognise the xcframework as something that could have a signature.

PS If we're going to discuss here, I guess its worth reopening this Issue.

**Comment by mungler** [comment]
The thing is, the patch to gen_xcframework.sh would be trivial if the resulting framework was something that the codesign tool saw as valid. But we need to figure out why it doesn't, I guess.

**Comment by mungler** [comment]
Actually, thinking about it a bit more, if - like us - you embed the JRE.xcframework in an app, you can get xcode to sign it as part of the app during build. So I think its perfectly valid for such a framework to be unsigned (though of course Apple encourage developers of frameworks to sign their output, for 'supply chain security' reasons) - the problem we have is more just that the package is not considered valid by the tool. So i'm not actually certain we need to bother patching the script to get it to sign the package as part of the j2objc build itself. 

TL;DR: I don't think we need to sign the package, it just needs to be in a format that the 'codesign' tool considers valid for querying and signing.

**Comment by bryan1anderson** [comment]
Firebase had something similar. Their fix might help us figure it out. https://github.com/firebase/firebase-ios-sdk/issues/11401

**Comment by mungler** [comment]
Well, i've actually solved this locally for my team, and i'm working on a patch for j2objc now. Namely moving the 'Headers' folder to the root of the bundle, instead of a symlink, and ditching the Versions folder entirely. With that setup, 'codesign' now correctly identifies the bundle as unsigned:

```
rory@Zaphod ~/W/j/d/frameworks (asmallworld/framework_patch)> ls JRE.xcframeworkHeaders/                        ios-arm64_x86_64-simulator/
Info.plist                      macos-arm64_x86_64/
ios-arm64_arm64e/               watchos-arm64_32_armv7k/
ios-arm64_x86_64-maccatalyst/   watchos-arm64_x86_64-simulator/
rory@Zaphod ~/W/j/d/frameworks (asmallworld/framework_patch)> codesign -dv JRE.xcframework/
JRE.xcframework/: code object is not signed at all
rory@Zaphod ~/W/j/d/frameworks (asmallworld/framework_patch) [1]> 
```

**Comment by bryan1anderson** [comment]
Yes that symlink and versions was the issue! @litstrong

**Comment by mungler** [comment]
https://github.com/google/j2objc/pull/2174/files

**Comment by mungler** [comment]
Fixed by PR https://github.com/google/j2objc/pull/2175

**Comment by litstrong** [comment]
Thanks @bryan1anderson and @mungler!

Looks like it is because Apple changes how they organize the folder structure (see [here](https://developer.apple.com/documentation/xcode/creating-a-multi-platform-binary-framework-bundle#Generate-the-XCFramework-bundle)). This can also be fixed by using command documented in the above to generate the XCFramework (I verified it can generate the same folder structure shown here, e.g.: without symlink on the Headers).

**Comment by mungler** [comment]
Fixed by PRs #2175 and #2181
